drivers/gpio/gpio_mcp23s17.c

```diff
@@ -0,0 +1,597 @@
+ * @param buf Buffer to read data into.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int _read_port_regs(struct device *dev, u8_t reg,
+			   union mcp23s17_port_data *buf)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	int ret;
+
+	u8_t addr = MCP23S17_ADDR | MCP23S17_READBIT;
+	u8_t buffer_tx[4] = { addr, reg, 0, 0 };
+
+	const struct spi_buf tx_buf = {
+		.buf = buffer_tx,
+		.len = 4,
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+	struct spi_buf rx_buf[2];
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = 2,
+	};
+	rx_buf[0].buf = NULL;
+	rx_buf[0].len = 1;
+	rx_buf[1].buf = &buf->byte;
+	rx_buf[1].len = 1;
+
+	ret = spi_transceive(drv_data->spi, &drv_data->spi_cfg, &tx, &rx);
+	if (ret) {
+		LOG_DBG("spi_transceive FAIL %d\n", ret);
+		return ret;
+	}
+
+	LOG_DBG("MCP23S17: Read: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
+		reg, buf->byte[0], (reg + 1), buf->byte[1]);
+
+	return 0;
+}
+
+/**
+ * @brief Write both port 0 and port 1 registers of certain register function.
+ *
+ * Given the register in reg, write the pair of port 0 and port 1.
+ *
+ * @param dev Device struct of the MCP23S17.
+ * @param reg Register to write into (the PORTA of the pair of registers).
+ * @param buf Buffer to write data from.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int _write_port_regs(struct device *dev, u8_t reg,
+			    union mcp23s17_port_data *buf)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	int ret;
+
+	u8_t addr = MCP23S17_ADDR;
+	u8_t buffer_tx[2] = { addr, reg };
+
+	const struct spi_buf tx_buf[2] = {
+		{
+			.buf = buffer_tx,
+			.len = 2,
+		},
+		{
+			.buf = &buf->byte,
+			.len = 2,
+		}
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = 2,
+	};
+
+	ret = spi_write(drv_data->spi, &drv_data->spi_cfg, &tx);
+	if (ret) {
+		LOG_DBG("spi_write FAIL %d\n", ret);
+		return ret;
+	}
+
+	LOG_DBG("MCP23S17: Write: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
+		reg, buf->byte[0], (reg + 1), buf->byte[1]);
+	return 0;
+}
+
+/**
+ * @brief Setup the pin direction (input or output)
+ *
+ * @param dev Device struct of the MCP23S17
+ * @param access_op Access operation (pin or port)
+ * @param pin The pin number
+ * @param flags Flags of pin or port
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int _setup_pin_dir(struct device *dev, int access_op,
+			  u32_t pin, int flags)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	union mcp23s17_port_data *port = &drv_data->reg_cache.iodir;
+	u16_t bit_mask;
+	u16_t new_value = 0;
+	int ret;
+
+	switch (access_op) {
+	case GPIO_ACCESS_BY_PIN:
+		bit_mask = 1 << pin;
+
+		/* Config 0 == output, 1 == input */
+		if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
```

Comment:
`GPIO_DIR_` are deprecated. The new API is using `GPIO_INPUT`, `GPIO_OUTPUT` flags.